### PR TITLE
PLAT-1231 | Users want to know if a UGC was verified

### PR DESCRIPTION
**Pull request by rogatty**
https://bethink.atlassian.net/browse/PLAT-1231

### CODE REVIEW

File: app/Models/Comment.php [hunk 1]
```diff
@@ -2,14 +2,16 @@
...
 use Illuminate\Database\Eloquent\SoftDeletes;
+use \Altek\Accountant\Recordable as RecordableTrait;
 
-class Comment extends Model implements WithReactions
+class Comment extends Model implements WithReactions, Recordable
 {
-	use SoftDeletes;
+	use SoftDeletes, RecordableTrait;
 
 	protected $fillable = ['text', 'user_id'];
```
Comment (rogatty) [inline app/Models/Comment.php]: I think we're missing `verified_at` here and `protected $dates = ['verified_at'];` below.

File: app/Policies/CommentPolicy.php [hunk 2]
```diff
@@ -12,7 +12,7 @@ class CommentPolicy
 
 	public function before($user, $ability)
 	{
-		if ($user->isAdmin()) {
+		if ($user->isAdmin() || $user->isModerator()) {
```
Comment (rogatty) [inline app/Policies/CommentPolicy.php]: I'd remove `$user->isAdmin() ||` from `update` and `delete` below as they're misleading now (moderator can do these actions too).

File: resources/assets/js/components/global/reactions/Vote.vue [hunk 3]
```diff
@@ -59,12 +50,12 @@
 </style>
 
 <script>
-import {
-	reaction
-} from 'js/mixins/reaction';
+import { reaction } from 'js/mixins/reaction';
+import WnlRectangleButton from 'js/components/RecatangleButton';
```
Comment (rogatty) [inline resources/assets/js/components/global/reactions/Vote.vue]: Typo in the filename.

File: resources/assets/js/components/UserGeneratedContentHeader.vue [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+						:can-verify="canVerify"
+						:resource="content"
+						@verify="$emit('verify')"
+						@unverify="$emit('unverify')"
+					/>
+				</span>
+			</div>
+		</div>
+		<wnl-modal v-if="modalVisible" @closeModal="closeModal">
+			<wnl-user-profile-modal :author="author" />
```
Comment (rogatty) [inline resources/assets/js/components/UserGeneratedContentHeader.vue]: It's `:profile=` in https://github.com/bethinkpl/wnl-platform/pull/1450, let's remember to update it in one of the branches.

File: resources/assets/js/components/comments/Comment.vue [hunk 5]
```diff
@@ -1,21 +1,17 @@
...
-					</p>
-				</div>
-			</figure>
+	<article class="wnl-comment">
+		<wnl-user-generated-content-header
+			class="wnl-comment__header"
+			:author="profile"
+			:content="comment"
+			:delete-target="target"
+			:delete-resource-rotue="requestRoute"
```
Comment (rogatty) [inline resources/assets/js/components/comments/Comment.vue]: Typo

File: resources/assets/js/components/comments/Comment.vue [hunk 6]
```diff
@@ -1,21 +1,17 @@
...
+			@deleteSuccess="$emit('deleteSuccess')"
+			@verify="$emit('verify')"
+			@unverify="$emit('unverify')"
+		/>
+		<div class="wrap content" v-html="comment.text" />
+		<div class="qna-answer__actions">
```
Comment (rogatty) [inline resources/assets/js/components/comments/Comment.vue]: `qna-answer`?

File: resources/assets/js/components/global/Verify.vue [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+		font-size: $font-size-minus-1
+		font-weight: bold
+		color: $color-gray
+		text-transform: uppercase
+		width: auto
+
+		&--verified
+			color: $color-green
+
+		&--can_verify
```
Comment (rogatty) [inline resources/assets/js/components/global/Verify.vue]: I believe this should be:
```suggestion
		&--can-verify
```

File: resources/assets/js/components/global/Verify.vue [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	</p>
+</template>
+
+<style lang="sass" rel="stylesheet/sass" scoped>
+	@import 'resources/assets/sass/variables'
+
+	.verify
+		font-size: $font-size-minus-1
+		font-weight: bold
+		color: $color-gray
```
Comment (rogatty) [inline resources/assets/js/components/global/Verify.vue]: 🔤

File: resources/assets/js/components/global/Verify.vue [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+<script>
+import FileVerifiedSvg from 'images/icons/file-verified.svg';
+
+export default {
+	components: {
+		FileVerifiedSvg,
+	},
+	props: {
+		resource: {
```
Comment (rogatty) [inline resources/assets/js/components/global/Verify.vue]: How do you feel about passing `resource.verified_at` instead? We would be explicit what is required, now you need to read the code to see that this should be an object with `verified_at` property.

File: resources/assets/js/components/qna/QnaQuestion.vue [hunk 10]
```diff
@@ -1,108 +1,71 @@
...
+					class="qna-question__actions__action"
+					type="up"
+					:reactable-id="questionId"
+					:reactable-resource="reactableResource"
+					:state="voteState"
+					module="qna"
+				/>
+				<template v-if="!readOnly">
+					<wnl-rectangle-button class="qna-question__actions__action" @click="showAnswerForm = !showAnswerForm">
+						<span>
```
Comment (rogatty) [inline resources/assets/js/components/qna/QnaQuestion.vue]: Do we need this wrapper?

File: resources/assets/js/components/qna/QnaQuestion.vue [hunk 11]
```diff
@@ -1,108 +1,71 @@
...
-					<a
-						v-if="showAnswerForm"
-						class="button is-small"
-						@click="showAnswerForm = false"
-					>
-						<span>Ukryj</span>
-					</a>
-				</div>
+		<transition name="fade">
+			<wnl-qna-new-answer-form
```
Comment (rogatty) [inline resources/assets/js/components/qna/QnaQuestion.vue]: Before the change, it was inside the container with `disabled` class, now it's not. Does it matter?

File: resources/assets/js/components/qna/QnaQuestion.vue [hunk 12]
```diff
@@ -127,55 +90,77 @@
...
-		strong
-			font-weight: $font-weight-black
+		&__content
+			font-size: $font-size-plus-1
+			justify-content: flex-start
+			padding-right: $margin-base
+			width: 100%
+			word-wrap: break-word
+			word-break: break-word
+			margin-bottom: $margin-big
```
Comment (rogatty) [inline resources/assets/js/components/qna/QnaQuestion.vue]: :abc:

File: resources/assets/js/components/qna/QnaQuestion.vue [hunk 13]
```diff
@@ -1,108 +1,71 @@
 <template>
 	<div class="qna-thread" :class="{'is-mobile': isMobile}">
```
Comment (rogatty) [inline resources/assets/js/components/qna/QnaQuestion.vue]: `is-mobile` class was removed.

File: resources/assets/js/components/qna/QnaQuestion.vue [hunk 14]
```diff
@@ -207,47 +188,31 @@
...
-import { timeFromS } from 'js/utils/time';
+import WnlRectangleButton from 'js/components/RecatangleButton';
 
 export default {
 	components: {
-		'wnl-delete': Delete,
-		'wnl-resolve': Resolve,
+		WnlUserGeneratedContentHeader,
+		WnlRectangleButton,
 		'wnl-vote': Vote,
```
Comment (rogatty) [inline resources/assets/js/components/qna/QnaQuestion.vue]: `WnlVote`?

File: resources/assets/js/components/qna/QnaQuestion.vue [hunk 15]
```diff
@@ -316,9 +276,6 @@ export default {
 		allAnswers() {
 			return this.answersFromHighestUpvoteCount;
 		},
-		tags() {
-			return this.questionTags(this.questionId).map((tag) => tag.name) || [];
```
Comment (rogatty) [inline resources/assets/js/components/qna/QnaQuestion.vue]: Why were the tags removed? They're still visible on the designs. I know we're deprecating them slowly but am not sure if they already have no use. Was it confirmed with @adamkarminski?

File: resources/assets/js/store/modules/comments.js [hunk 16]
```diff
@@ -171,13 +157,73 @@ export const commentsActions = {
...
+			await _updateComment(payload.id, { resolved: true });
+			commit(types.UPDATE_COMMENT, {
+				id: payload.id,
+				resource: payload.commentableResource,
+				resourceId: payload.commentableId,
+				updatedComment: { resolved: true }
+			});
+		} catch (e) {
+			$wnl.logger.error(e);
+			dispatch('addAutoDismissableAlert', {
```
Comment (rogatty) [inline resources/assets/js/store/modules/comments.js]: I'd avoid touching UI in the store if possible.

File: resources/assets/js/store/modules/comments.js [hunk 17]
```diff
@@ -171,13 +157,73 @@ export const commentsActions = {
...
+				text: 'Ups, coś poszło nie tak. Spróbuj ponownie, a jeżeli to nie pomoże to daj nam znać o błędzie.',
```
Comment (rogatty) [inline resources/assets/js/store/modules/comments.js]: Can we use a `ui.error.tryAgainOrReport` translation? Would be easier to keep them all consistent.

File: resources/assets/js/store/modules/comments.js [hunk 18]
```diff
@@ -171,13 +157,73 @@ export const commentsActions = {
...
-		_resolveComment(payload.id, false)
-			.then(() => commit(types.UNRESOLVE_COMMENT, payload));
+	async unresolveComment({ commit, dispatch }, payload) {
+		try {
+			await _updateComment(payload.id, { resolved: false });
+			commit(types.UPDATE_COMMENT, {
+				id: payload.id,
+				resource: payload.commentableResource,
+				resourceId: payload.commentableId,
+				updatedComment: { resolved: false }
```
Comment (rogatty) [inline resources/assets/js/store/modules/comments.js]: Why is it different from `verifyComment` where we get the updated comment data from `_updateComment`?

File: resources/assets/js/store/modules/qna.js [hunk 19]
```diff
@@ -333,13 +336,53 @@ const actions = {
...
+			$wnl.logger.error(e);
+			dispatch('addAutoDismissableAlert', {
+				text: 'Ups, coś poszło nie tak. Spróbuj ponownie, a jeżeli to nie pomoże to daj nam znać o błędzie.',
+				type: ALERT_TYPES.ERROR,
+			}, { root: true });
+		}
+	},
+	async unresolveQuestion({ commit, dispatch }, questionId) {
+		try {
+			await _updateQuestion(questionId, { resolved: false });
```
Comment (rogatty) [inline resources/assets/js/store/modules/qna.js]: Why is it different from `verifyQuestion ` where we get the updated comment data from `_updateQuestion `?

File: resources/assets/js/store/modules/qna.js [hunk 20]
```diff
@@ -333,13 +336,53 @@ const actions = {
...
+			commit(types.QNA_UNRESOLVE_QUESTION, { questionId });
```
Comment (rogatty) [inline resources/assets/js/store/modules/qna.js]: How about using `QNA_UPDATE_QUESTION` for consistency?

File: tests/Api/Comments/CommentsTest.php [hunk 21]
```diff
@@ -68,4 +69,35 @@ public function delete_comment()
 			->assertStatus(200);
 	}
 
+	public function testCommentVerifyUnverify()
```
Comment (rogatty) [inline tests/Api/Comments/CommentsTest.php]: How about checking 403 for all resources?

File: tests/Api/Comments/CommentsTest.php [hunk 22]
```diff
@@ -68,4 +69,35 @@ public function delete_comment()
...
+	{
+		Comment::flushEventListeners();
+		QnaQuestion::flushEventListeners();
+		QnaAnswer::flushEventListeners();
```
Comment (rogatty) [inline tests/Api/Comments/CommentsTest.php]: Do we need to do that if the comment isn't related to a QnA?

File: tests/Api/Qna/AnswersTest.php [hunk 23]
```diff
@@ -52,4 +53,35 @@ public function update_answer()
 		$response
 			->assertStatus(200);
 	}
+
+	public function testQnaAnswerVerifyUnverify()
+	{
+		QnaAnswer::flushEventListeners();
+		QnaQuestion::flushEventListeners();
```
Comment (rogatty) [inline tests/Api/Qna/AnswersTest.php]: Do we need to do that if the answer isn't related to a question?

File: webpack.mix.js [hunk 24]
```diff
@@ -30,6 +30,20 @@ if (mix.inProduction()) {
 	mix.version();
 }
 
+// Disable loading svgs by file-loaderto be able to use vue-svg-loader
+Mix.listen('configReady', function(config) {
+	const rules = config.module.rules;
+	const targetRe = /(\.(png|jpe?g|gif|webp)$)/;
+	const originalRegex = /(\.(png|jpe?g|gif|webp)$|^((?!font).)*\.svg$)/;
```
Comment (rogatty) [inline webpack.mix.js]: Won't this break `src="{{ asset('images/star-o.svg') }}"`?

File: resources/assets/js/components/global/Verify.vue (hunk 8 above)
Comment (bognix) [reply resources/assets/js/components/global/Verify.vue]: you are talking about alphabetical order? :)

File: resources/assets/js/components/qna/QnaQuestion.vue (hunk 10 above)
Comment (bognix) [reply resources/assets/js/components/qna/QnaQuestion.vue]: Yes, I stumbled upon some problems when I tried to passed more than one root element to slot.

File: resources/assets/js/components/qna/QnaQuestion.vue (hunk 15 above)
Comment (bognix) [reply resources/assets/js/components/qna/QnaQuestion.vue]: We don't display them next to the QnaQuestion. They were displayed only on the discussions visible on Dashboard but this was removed recently. We don't fetch the qna with tags any more

Comment (rogatty) [reply resources/assets/js/components/qna/QnaQuestion.vue]: Ah, I got it confused with `contextTags` in `Qna.vue` 👍

File: resources/assets/js/components/global/Verify.vue (hunk 8 above)
Comment (rogatty) [reply resources/assets/js/components/global/Verify.vue]: Yup.

File: resources/assets/js/components/qna/QnaQuestion.vue (hunk 11 above)
Comment (rogatty) [reply resources/assets/js/components/qna/QnaQuestion.vue]: Yeah, it matters. Now you can send an answer to a resolved question and there is an error:
```
Ups, zapisywanie nie powiodło się. Najprawdopodobniej treść do której się odwołujesz została usunięta. W przypadku dalszych problemów odśwież stronę.
```

File: resources/assets/js/store/modules/comments.js (hunk 18 above)
Comment (kvas-damian) [reply resources/assets/js/store/modules/comments.js]: Resolving isn't a normal action. API doesn't return `resolved` flag. Backend is soft-deleting the entity and doesn't return it anymore. On the frontend we allow a moderator to revert the action until the list is loaded again. That's why we set the flag here.

File: resources/assets/js/store/modules/qna.js (hunk 19 above)
Comment (kvas-damian) [reply resources/assets/js/store/modules/qna.js]: see https://github.com/bethinkpl/wnl-platform/pull/1455#discussion_r279508671

File: resources/assets/js/store/modules/qna.js (hunk 20 above)
Comment (kvas-damian) [reply resources/assets/js/store/modules/qna.js]: I was there. I didn't dig into details but vuex wasn't updating after attribute creation. That's why I left them as separate mutations.

File: webpack.mix.js (hunk 24 above)
Comment (kvas-damian) [reply webpack.mix.js]: It looks like a `blade` feature and it works.

File: resources/assets/js/tests/setup.js [hunk 25]
```diff
@@ -0,0 +1,8 @@
+global.$wnl = {
+	env: {
+		appUrl: 'http://test'
+	},
+	logger: {
```
Comment (bognix) [inline resources/assets/js/tests/setup.js]: remove logger from here.

Comment (rogatty) [reply resources/assets/js/tests/setup.js]: Why? Where would you put it?

File: svgTransformer.js [hunk 26]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
```
Comment (rogatty) [inline svgTransformer.js]: This file is redundant, correct?

Comment (bognix) [reply svgTransformer.js]: No it's not. It's used in `jest.config.js`

File: resources/assets/js/components/UserGeneratedContentHeader.vue [hunk 27]
```diff
@@ -1,90 +1,90 @@
...
-			:full-name="author.full_name"
-			:url="author.avatar"
-			size="medium"
-			@click.native="showModal"
-		/>
-		<div class="user-generated-content-header__info">
-			<span :class="{'author-forgotten': author.deleted_at}" @click="showModal">
-				{{author.full_name}}
-			</span>
+		<div class="user-generated-content-header__avatar">
```
Comment (rogatty) [inline resources/assets/js/components/UserGeneratedContentHeader.vue]: This div contains more than an avatar.

File: resources/assets/js/components/UserGeneratedContentHeader.vue [hunk 28]
```diff
@@ -1,90 +1,90 @@
...
-					<wnl-verify
-						:can-verify="canVerify"
-						:verified-at="content.verified_at"
-						@verify="$emit('verify')"
-						@unverify="$emit('unverify')"
-					/>
-				</span>
 			</div>
 		</div>
+		<span class="user-generated-content-header__meta__item">
```
Comment (rogatty) [inline resources/assets/js/components/UserGeneratedContentHeader.vue]: It's outside of `user-generated-content-header__meta`. If this is correct, then the name is confusing.

File: resources/assets/js/components/UserGeneratedContentHeader.vue [hunk 29]
```diff
@@ -1,90 +1,90 @@
...
+					{{author.full_name}}
+				</span>
+				<span class="user-generated-content-header__separator">·</span>
 				<span class="user-generated-content-header__meta__item">
-					<span class="user-generated-content-header__separator">·</span>
 					<span>{{time}}</span>
 				</span>
-				<span v-if="canDelete" class="user-generated-content-header__meta__item">
-					<span class="user-generated-content-header__separator">·</span>
+				<span v-if="hasMeta" class="user-generated-content-header__separator">·</span>
```
Comment (rogatty) [inline resources/assets/js/components/UserGeneratedContentHeader.vue]: `user-generated-content-header__meta` is a parent so `hasMeta` is misleading.

File: resources/assets/js/components/global/Delete.vue [hunk 30]
```diff
@@ -29,7 +38,24 @@ import { ALERT_TYPES } from 'js/consts/alert';
...
 	name: 'Delete',
-	props: ['target', 'requestRoute', 'title'],
+	props: {
+		target: {
+			type: String,
+			default: ''
+		},
+		requestRoute: {
+			type: String,
+			default: '',
```
Comment (rogatty) [inline resources/assets/js/components/global/Delete.vue]: This component doesn't make sense without `requestRoute` so I'd update it to `required: true`

File: resources/assets/js/components/global/Delete.vue [hunk 31]
```diff
@@ -29,7 +38,24 @@ import { ALERT_TYPES } from 'js/consts/alert';
...
+		},
+		title: {
+			type: String,
+			default: ''
```
Comment (rogatty) [inline resources/assets/js/components/global/Delete.vue]: Make it `'Usuń'` and we can remove `this.computedTitle`

File: resources/assets/js/components/global/Delete.vue [hunk 32]
```diff
@@ -29,7 +38,24 @@ import { ALERT_TYPES } from 'js/consts/alert';
 
 export default {
 	name: 'Delete',
-	props: ['target', 'requestRoute', 'title'],
+	props: {
+		target: {
+			type: String,
+			default: ''
```
Comment (rogatty) [inline resources/assets/js/components/global/Delete.vue]: `Chcesz usunąć ?` doesn't make sense.

File: resources/assets/js/components/global/Resolve.vue [hunk 33]
```diff
@@ -1,33 +1,51 @@
 <template>
-	<p
-		v-if="resource.resolved"
-		class="resolvable icon"
+	<div
+		:class="['resolvable', touchable && 'resolvable--touchable']"
 		@click="$emit('unresolveResource')"
```
Comment (rogatty) [inline resources/assets/js/components/global/Resolve.vue]: This should be emitted on undo element.

File: resources/assets/js/components/global/Resolve.vue [hunk 34]
```diff
@@ -1,33 +1,51 @@
...
-		<i class="fa fa-undo" />
-		<span v-t="'ui.action.undo'" />
-	</p>
-	<span v-else class="resolvable icon">
-		<i class="fa fa-check" @click="$emit('resolveResource')" />
-	</span>
+		<div
+			v-if="resource.resolved"
+			class="icon"
+			title="Rozwiąż"
```
Comment (rogatty) [inline resources/assets/js/components/global/Resolve.vue]: You can resolve when it's resolved? This template is messed up.

File: resources/assets/js/components/global/Delete.vue [hunk 35]
```diff
@@ -1,20 +1,29 @@
...
 
-	.icon
+	.wnl-delete
 		cursor: pointer
 
+		&--touchable
+			min-width: $min-touchable-size
+			min-height: $min-touchable-size
+			display: flex
+			align-items: center
```
Comment (kvas-damian) [inline resources/assets/js/components/global/Delete.vue]: I'm not sure why `touchable` modifier is mixed with centering

File: resources/assets/js/components/global/reactions/Watch.vue [hunk 36]
```diff
@@ -16,7 +16,11 @@
 	@import 'resources/assets/sass/variables'
 
 	.watch
-		color: $color-gray
+		font-weight: 600
```
Comment (kvas-damian) [inline resources/assets/js/components/global/reactions/Watch.vue]: I believe we don't support `600` in our fonts

File: resources/assets/js/components/qna/QnaQuestion.vue [hunk 37]
```diff
@@ -147,11 +149,25 @@
...
 			&__action
-				margin-right: $margin-base
-				.icon.answer-icon
-					margin-left: $margin-small
+				margin-top: $space-xxs
+				margin-right: $space-s
+
+				&.qna-question__answer
+					color: $color-background-gray
+					font-weight: 600
```
Comment (kvas-damian) [inline resources/assets/js/components/qna/QnaQuestion.vue]: we don't support 600

File: resources/assets/js/components/global/Delete.vue (hunk 35 above)
Comment (bognix) [reply resources/assets/js/components/global/Delete.vue]: to make sure content inside is centered - the icon.

File: resources/assets/js/components/global/Resolve.vue (hunk 34 above)
Comment (bognix) [reply resources/assets/js/components/global/Resolve.vue]: only titles where messed up.

File: resources/assets/js/components/global/Resolve.vue (hunk 33 above)
Comment (rogatty) [reply resources/assets/js/components/global/Resolve.vue]: Fixed in https://github.com/bethinkpl/wnl-platform/commit/7e6b985c0baf43aafff09e3248800eb878890446

**Pull request merged**